### PR TITLE
etcdctl/ctlv2: total-timeout for Sync

### PR DESCRIPTION
**Pull request by gyuho**
Fix https://github.com/coreos/etcd/issues/4897.

### CODE REVIEW

**Comment by gyuho** [comment]
/cc @xiang90 @kayrus 

Please review.

Thanks.

File: etcdctl/ctlv2/command/util.go [hunk 1]
```diff
@@ -156,7 +156,12 @@ func getTransport(c *cli.Context) (*http.Transport, error) {
 		CertFile: certfile,
 		KeyFile:  keyfile,
 	}
-	return transport.NewTransport(tls, defaultDialTimeout)
+
+	to := c.GlobalDuration("total-timeout")
```
Comment (xiang90) [inline etcdctl/ctlv2/command/util.go]: why shall we do this? say if the total-timeout is 10 seconds and the defaultDialTimeout is 5 seconds, the change might cause the dailtimeout to be rewrote by a longer one which is 10 seconds. this is incorrect.

**Comment by gyuho** [comment]
@xiang90 You are right. Just reverted it.

PTAL.

Thanks.

File: etcdctl/ctlv2/command/util.go (hunk 1 above)
Comment (xiang90) [inline etcdctl/ctlv2/command/util.go]: if totalTimeout !=0 && to < defaultDialTimeout {
    dialTimeout = totalTimeout
}

/cc @gyuho

File: etcdctl/ctlv2/command/util.go [hunk 2]
```diff
@@ -203,7 +203,7 @@ func mustNewClient(c *cli.Context) client.Client {
 		if debug {
 			fmt.Fprintf(os.Stderr, "start to sync cluster using endpoints(%s)\n", strings.Join(hc.Endpoints(), ","))
 		}
-		ctx, cancel := context.WithTimeout(context.Background(), client.DefaultRequestTimeout)
+		ctx, cancel := contextWithTotalTimeout(c)
```
Comment (xiang90) [inline etcdctl/ctlv2/command/util.go]: also here, we should also add a timeout for the request timeout.

File: etcdctl/ctlv2/command/util.go (hunk 1 above)
Comment (gyuho) [inline etcdctl/ctlv2/command/util.go]: @xiang90 Did you mean

```go
	to := c.GlobalDuration("total-timeout")
	if to != 0 && to > defaultDialTimeout {
		defaultDialTimeout = to
	}
	return transport.NewTransport(tls, defaultDialTimeout)
```

Comment (gyuho) [inline etcdctl/ctlv2/command/util.go]: Oh I meant:

```go
	to := c.GlobalDuration("total-timeout")
	if to != 0 && to < defaultDialTimeout {
		defaultDialTimeout = to
	}
	return transport.NewTransport(tls, defaultDialTimeout)
```

Comment (xiang90) [inline etcdctl/ctlv2/command/util.go]: no. 
Your code has two issues:

1. we should never modify the default variable in any case.
2. we should keep the shorter timeout not longer. Say I want the total timeout to be 5 seconds, but the defaultDialtimeout is 30 seconds. I should give the transport a timeout of 5 seconds, not 30 seconds. Or dial might block for 30 seconds, which is longer than 5 seconds.

```go
    dailTimeout := defaultDialTimeout 
    totalTimeout := c.GlobalDuration("total-timeout")
    if totalTimoeout != 0 && totalTimeout < defaultDialTimeout {
        defaultDialTimeout = totalTimeout
    }
    return transport.NewTransport(tls, defaultDialTimeout)
```

File: etcdctl/ctlv2/command/util.go (hunk 2 above)
Comment (gyuho) [inline etcdctl/ctlv2/command/util.go]: @xiang90 

> also here, we should also add a timeout for the request timeout.

Sorry, can you explain more? This already puts timeout for the command:

```go
func contextWithTotalTimeout(c *cli.Context) (context.Context, context.CancelFunc) {
	return context.WithTimeout(context.Background(), c.GlobalDuration("total-timeout"))
}
``` 

`c.GlobalDuration("timeout")` is already used for `HeaderTimeoutPerRequest` here https://github.com/coreos/etcd/blob/master/etcdctl/ctlv2/command/util.go#L302

Comment (xiang90) [inline etcdctl/ctlv2/command/util.go]: Oh. I am not aware of https://github.com/coreos/etcd/blob/master/etcdctl/ctlv2/command/util.go#L302.

Then this is fine. Thanks.

**Comment by xiang90** [comment]
Can you give it a try after the change? Does this fix the problem mentioned in #4897

**Comment by gyuho** [comment]
@xiang90 Yeah, just pushed a change to address your last comment and tried

```
./bin/etcdctl --debug --timeout 5s --endpoint http://example.com:1234 ls /

start to sync cluster using endpoints(http://example.com:1234)
cURL Command: curl -X GET http://example.com:1234/v2/members
Error:  client: etcd cluster is unavailable or misconfigured
error #0: dial tcp 93.184.216.34:1234: i/o timeout
```

**Comment by xiang90** [comment]
lgtm

**Comment by gyuho** [comment]
Thanks. Merging in greens.

**Pull request merged**